cupy/cuda/cupy_nccl.h

```diff
@@ -35,11 +43,103 @@ void ncclCommDestroy(...);
 ncclResult_t ncclCommCuDevice(...);
 ncclResult_t ncclCommUserRank(...);
 ncclResult_t ncclAllReduce(...);
-ncclResult_t  ncclReduce(...);
-ncclResult_t  ncclBcast(...);
+ncclResult_t ncclReduce(...);
+ncclResult_t ncclBcast(...);
+ncclResult_t ncclReduceScatter(...);
+ncclResult_t ncclAllGather(...);
+
+typedef struct CUstream_st *cudaStream_t;
 
 }
 
 #endif // #ifndef CUPY_NO_CUDA
 
+#ifndef NCCL_MAJOR
+#define NCCL_MAJOR 1
+#define NCCL_MINOR 0
+#define NCCL_PATCH 0
+#endif
+
+#define NCCL_VERSION  (NCCL_MAJOR * 1000 + NCCL_MINOR * 100 + NCCL_PATCH)
+
+
+#if (NCCL_VERSION >= 2000)
+
+ncclDataType_t _get_proper_datatype(ncclDataType_t datatype) {
+    return datatype;
+}
+
+#else // #if (NCCL_VERSION >= 2000)
+
+#define NCCL_CHAR_V1 ncclChar
+#define NCCL_INT_V1 ncclInt
+#define NCCL_HALF_V1 ncclHalf
```

Comment:
`nccl.h` does not declare `ncclHalf` on no half environment.
```
#ifdef CUDA_HAS_HALF
               ncclHalf       = 2,
#endif
```
Cloud you fix?

Reply:
It looks the issue happens when CUDA 7.0 or older, is it correct?
Will check and fix it.

Reply:
I think I reproduced the issue. The followings are part of message when the command `python setup.py develop` is run with CUDA7 and NCCL1.

    $ python setup.py develop
    Options: {'annotate': False, 'linetrace': False, 'profile': False, 'no_cuda': False}
    Include directories: ['/usr/local/cuda-7.0/include']
    Library directories: ['/usr/local/cuda-7.0/lib64', '/usr/local/cuda-7.0/lib']
    ...
    In file included from cupy/cuda/nccl.cpp:482:0:
    cupy/cuda/cupy_nccl.h:76:22: error: ncclHalf was not declared in this scope
     #define NCCL_HALF_V1 ncclHalf
                          ^
    cupy/cuda/cupy_nccl.h:90:5: note: in expansion of macro NCCL_HALF_V1
         NCCL_HALF_V1,         // ncclFloat16, ncclHalf
         ^
    error: command 'gcc' failed with exit status 1

Reply:
BTW, what CUDA versions will be supported by CuPy 2? I would like to know if CUDA 7.0 will be included in support versions.